### PR TITLE
Fix libunwind compilation with GCC 10

### PR DESCRIPTION
**Pull request by nickrobison-usds**
Compilation is failing on GCC 10 due to changes in default flags.

Apply the patch detailed here: https://github.com/libunwind/libunwind/pull/166 when building on GCC 10+.

### CODE REVIEW

**Review by adamjstewart — APPROVED** [review]

File: var/spack/repos/builtin/packages/libunwind/package.py [hunk 1]
```diff
@@ -32,6 +32,8 @@ class Libunwind(AutotoolsPackage):
             description='Support zlib compressed symbol tables '
             '(1.5 and later).')
 
+    patch('gcc.patch', when=':@1.6 %gcc@10.0:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/libunwind/package.py]:
```suggestion
    patch('gcc.patch', when='@:1.6 %gcc@10.0:')
```
This fixes the failing unit tests.

Comment (michaelkuhn) [reply var/spack/repos/builtin/packages/libunwind/package.py]: Apart from this, the patch only applies cleanly to 1.4 (the older versions fail and 1.5 already has it applied). I would suggest to simply do something like this (so we do not have to carry the large patch):
```python
def setup_build_environment(self, env):
    # https://github.com/libunwind/libunwind/pull/166
    if self.spec.satisfies('%gcc@10:'):
        env.set('CFLAGS', '-fcommon')
```

**Review by michaelkuhn — CHANGES_REQUESTED** [review]

**Review by michaelkuhn — APPROVED** [review]

**Comment by mwkrentel** [comment]
Am I missing something obvious here? 
Did y'all actually try building this with gcc 10.x?

I tried adding this (without the conditional)

```
    def setup_build_environment(self):
        env.set('CFLAGS', '-fcommon')
```

And spack install crashed with:

```
==> Error: TypeError: setup_build_environment() takes 1 positional argument but 2 were given
```

**Comment by michaelkuhn** [comment]
>     def setup_build_environment(self):
>         env.set('CFLAGS', '-fcommon')

You are missing the second `env` parameter.

Edit: Oops, sorry, I just noticed that this is missing in @nickrobison-usds's version as well (but was included in my suggestion). @nickrobison-usds please fix this. Thanks!

File: var/spack/repos/builtin/packages/libunwind/package.py [hunk 2]
```diff
@@ -49,6 +49,11 @@ class Libunwind(AutotoolsPackage):
 
     flag_handler = AutotoolsPackage.build_system_flags
 
+    def setup_build_environment(self):
```
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/libunwind/package.py:52]:
```suggestion
    def setup_build_environment(self, env):
```

**Comment by mwkrentel** [comment]
There is another issue, one that I'm very reluctant to bring up.

By setting `env.set('CFLAGS', '-fcommon')`, configure will see this in
the environment and use exactly `CFLAGS=-fcommon` and drop the `-g -O2`.
That is, spack will build libunwind without optimization.  For a longer
description, see #14430.

Spack doesn't have the tools to handle this cleanly.  You can see how
I worked around this (awkwardly) in libmonitor and intel-xed.

If you like, you can ignore this for now and I'll do something the
next time I add a new version.

And yes, please fix the `self, env` typo/bug.

**Comment by adamjstewart** [comment]
@mwkrentel what about `env.append_flags('CFLAGS', '-fcommon')`?

**Comment by michaelkuhn** [comment]
I think the problem is that autotools will behave differently depending on whether CFLAGS is set or not. If it is not set at all (which is the case without this modification), autotools will use a sane default. If it is set to anything, it will just use this value.

We could also use `flag_handler`, which seems to set `SPACK_CFLAGS`, which is then set internally by the compiler wrappers.

**Comment by michaelkuhn** [comment]
Just gave this a try. The following also works:
```diff
-    flag_handler = AutotoolsPackage.build_system_flags
+    def flag_handler(self, name, flags):
+        if name == 'cflags':
+            if self.spec.satisfies('@:1.4 %gcc@10:'):
+                flags.append('-fcommon')
+
+        return (flags, None, None)
```
But I am not 100% sure what `flag_handler = AutotoolsPackage.build_system_flags` does and whether this will break something else.

Edit: To clarify, this will set `SPACK_CFLAGS`, which autotools does not pick up and we still get the sane default of `-g -O2`.

**Comment by mwkrentel** [comment]
Yes, the basic problem is that unset and set to empty are very
different in autotools world.

Note: the triple returned from `flag_handler` is: (1) at the front of
the spack wrapper (inject flags), (2) in the environment, and (3) on
the configure line (build system).  So, `build_system_flags` is the
same as `return (None, None, flags)`.

The problem with putting the flags in the inject flags is that
configure doesn't see them, so autotools will then choose `-g -O2`.
For `-fcommon`, this doesn't matter, but if you wrote `cflags='-O0'`,
then you'd end up with `-O0 -g -O2`.  That's less of a sin, but still
not correct.

I think it may work to put `-fcommon` in inject flags and show the
rest to configure.  Something like:

```
if ...:
    return ('-fcommon', None, flags)
```

Btw, IMO, it's not really necessary to make old versions run with the
newest compilers, especially if that compiler didn't exist at the time
the version was released.  If it already works out of the box for
1.5-rc1, then if it were me, I'd just do nothing, or else add a conflict.

**Comment by michaelkuhn** [comment]
> Btw, IMO, it's not really necessary to make old versions run with the
> newest compilers, especially if that compiler didn't exist at the time
> the version was released. If it already works out of the box for
> 1.5-rc1, then if it were me, I'd just do nothing, or else add a conflict.

Agreed, but since we prefer version 1.4.0 at the moment, that makes for a bad out-of-the-box experience in my opinion. I think at least the default versions should compile with the current compiler so users can just do `spack install something` without having to figure out that they should manually specify a newer version of a certain package.

**Comment by mwkrentel** [comment]
Ok, fair enough.  I forgot that we don't count 1.5-rc1 as a "real"
release, even though I'm the one that set preferred.  :-)

**Comment by michaelkuhn** [comment]
I've verified that the following seems to do what we want:
```diff
-    flag_handler = AutotoolsPackage.build_system_flags
+    def flag_handler(self, name, flags):
+        wrapper_flags = None
+
+        if name == 'cflags':
+            # https://github.com/libunwind/libunwind/pull/166
+            if self.spec.satisfies('@:1.4 %gcc@10:'):
+                wrapper_flags = ['-fcommon']
+
+        return (wrapper_flags, None, flags)
```

@nickrobison-usds Sorry for hijacking this PR. :smile:

**Comment by mwkrentel** [comment]
Yes, I believe that should work.  I approve.
I think this is a good idiom for handling this situation,
at least for the state of spack today.

@michaelkuhn It's up to you, but if you want to add yourself as
another maintainer, feel free.

**Comment by michaelkuhn** [comment]
@nickrobison-usds Do you want to update this PR or should I just open a new one with the updated fix?

Edit: Nevermind, I have opened #16782 since more bug reports are coming in (#16780).

**Pull request closed without merging**